gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMethodDescriptor.java

```diff
@@ -30,8 +30,187 @@
 package com.google.api.gax.httpjson;
 
 import com.google.api.core.BetaApi;
+import com.google.auto.value.AutoValue;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import java.io.IOException;
+import java.io.Reader;
+import java.lang.reflect.Type;
+import java.util.Set;
 
 @BetaApi
-public class ApiMethodDescriptor<RequestT, ResponseT> {
-  // TODO implement
+@AutoValue
+/* Method descriptor for messages to be transmitted over HTTP. */
+public abstract class ApiMethodDescriptor<RequestT, ResponseT> {
+  public abstract String getFullMethodName();
+
+  public abstract Gson getBaseGson();
+
+  public abstract Gson getRequestMarshaller();
+
+  public abstract Gson getResponseMarshaller();
+
+  public abstract Type getRequestType();
+
+  public abstract Type getResponseType();
+
+  public abstract Set<String> getPathParams();
+
+  public abstract Set<String> getQueryParams();
+
+  public abstract String getHttpMethod();
+
+  public abstract HttpRequestFormatter getHttpRequestBuilder();
+
+  /* In the form "[prefix]%s[suffix]", where
+   *    [prefix] is any string; if length greater than 0, it should end with '/'.
+   *    [suffix] is any string; if length greater than 0, it should begin with '/'.
+   * This String format is applied to a serialized ResourceName to create the relative endpoint path.
+   */
+  public abstract String endpointPathTemplate();
+
+  public static <RequestT, ResponseT> ApiMethodDescriptor<RequestT, ResponseT> create(
+      String fullMethodName,
+      RequestT requestInstance,
+      ResponseT responseInstance,
+      String endpointPathTemplate,
+      Set<String> pathParams,
+      Set<String> queryParams,
+      HttpRequestFormatter httpRequestFormatter,
+      String httpMethod) {
```

Comment:
There are a lot of parameters here - we should go through the builder, and set any values to defaults that we can construct

Reply:
I put default values for the fields in newBuilder(). Let me know if you meant something else?

Reply:
What I mean is that we shouldn't have a non-private method that takes so many parameters. Users can't use both this method (`create`) and `newBuilder`.

Reply:
Ok, the create() method is private.